### PR TITLE
8 dummy pipeline

### PR DESCRIPTION
**Pull request by Wanille**
- [x] Dummy matcher
- [x] Dummy throttler
- [x] Dummy labeling function

### CODE REVIEW

**Review by macmacpri — APPROVED** [review]

**Review by thuettel — APPROVED** [review]

File: apply_lf.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+from fonduer.supervision import Labeler
+from fonduer.parser.models import Document
+
+from pipeline.utils import get_session
+from definitions.candidates import NameFullAbbr, NameFullTask
+from pipeline.lfs.dummy_lfs import dummy_lfs
+
+
+def main(db_name):
+    PARALLEL = 12
```
Comment (Wanille) [inline apply_lf.py:10]: use `from pipeline.utils import PARALLEL`

File: data_extractor.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+
+# imports
+import os, shutil
+
+## Wichtig: bei mir funktioniert der Dev-Container nicht, deshalb konnte nicht bisher getestet werden
```
Comment (Wanille) [inline data_extractor.py:5]: Wurde getestet Kommentar kann gelöscht werden

File: apply_lf.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+from fonduer.supervision import Labeler
+from fonduer.parser.models import Document
+
+from pipeline.utils import get_session
+from definitions.candidates import NameFullAbbr, NameFullTask
+from pipeline.lfs.dummy_lfs import dummy_lfs
+
+
+def main(db_name):
```
Comment (Wanille) [inline apply_lf.py:9]:
```suggestion
def main(db_name: str):
```

File: doc_parser.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
+from pipeline.utils import get_session
+from fonduer.meta import Meta
+from fonduer.parser.preprocessors import HTMLDocPreprocessor
+from fonduer.parser import Parser
+
+
+def main(db_name):
+    PARALLEL = 12
```
Comment (Wanille) [inline doc_parser.py:8]: See above. remove this line

File: pipeline/machine_learning/additional_notes_regression_model copy.py
Comment (Wanille) [inline pipeline/machine_learning/additional_notes_regression_model copy.py:1]: remove this file. new file `simple_model.py` contains this

File: pipeline/machine_learning/additional_notes_simple_model.py
Comment (Wanille) [inline pipeline/machine_learning/additional_notes_simple_model.py:1]: remove this file. New File `simple_model.py` contains all of this

File: pipeline/machine_learning/custom_get_candidates.py
Comment (Wanille) [inline pipeline/machine_learning/custom_get_candidates.py:1]: Move the content of this file into `pipeline.utils.py`

File: pipeline/machine_learning/regression_model.py
Comment (Wanille) [inline pipeline/machine_learning/regression_model.py:1]: Remove this file. New file `simple_model.py` contains this file.

File: pipeline/machine_learning/simple_model_old.py
Comment (Wanille) [inline pipeline/machine_learning/simple_model_old.py:1]: Remove this File. `simple_model.py` contains this.

File: pipeline/throttler/dummythrottler.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+from fonduer.utils.data_model_utils.tabular import same_table
+
+def throt_same_table(c):
```
Comment (Wanille) [inline pipeline/throttler/dummythrottler.py:3]: Do we have any mentions inside of tables?

File: simple_model.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+from emmental.data import EmmentalDataLoader
+from fonduer.learning.dataset import FonduerDataset
+import numpy as np
+from emmental.model import EmmentalModel
+from fonduer.learning.task import create_task
+from emmental.learner import EmmentalLearner
+
+
+def main(db_name):
+    PARALLEL = 12
```
Comment (Wanille) [inline simple_model.py:28]: See above. Remove this line

File: pipeline/machine_learning/custom_get_candidates.py
Comment (Wanille) [inline pipeline/machine_learning/custom_get_candidates.py:1]: Maybe rename to load_candidates

File: simple_model.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+    ).lf_summary() #Y=L_gold_train[0].reshape(-1)) #adds empirical accuracy to statistics
+
+
+    # Training a simple model based on the Labelfunctions to estimate their accuracies
+    gen_model = LabelModel(cardinality=2) # <-- cardinality = number of classes
+    gen_model.fit(L_train[0], n_epochs=500, log_freq=100)
+
+
+    # We now apply the generative model to the training candidates to get the noise-aware training label set. We'll refer to these as the training marginals:
+    train_marginals = gen_model.predict_proba(L_train[0]) # <-- TH: do not understand why L_train instead of train_cands, but backed up by two notebooks of Fonduer
```
Comment (Wanille) [inline simple_model.py:59]: save this somewhere so that we can analyze them like the wiki_president example does.

File: simple_model.py [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+    labeler.apply(split=1, lfs=[dummy_lfs, dummy_lfs], parallelism=PARALLEL)
+    L_dev = labeler.get_label_matrices(dev_cands)
+    print(L_dev[0].shape)
+
+    # We now could check the F1 score, but most likely our model is overfitted to the data and therefore it will likely be too good
+
+    
+
+    # Setup training config
+    config_cheese = {
```
Comment (Wanille) [inline simple_model.py:71]: maybe move config into `pipeline.utils.py` or create seperate json file with config.

File: simple_model.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+    gen_model.fit(L_train[0], n_epochs=500, log_freq=100)
+
+
+    # We now apply the generative model to the training candidates to get the noise-aware training label set. We'll refer to these as the training marginals:
+    train_marginals = gen_model.predict_proba(L_train[0]) # <-- TH: do not understand why L_train instead of train_cands, but backed up by two notebooks of Fonduer
+
+    # Apply the generative model to the development data set
+    labeler.apply(split=1, lfs=[dummy_lfs, dummy_lfs], parallelism=PARALLEL)
+    L_dev = labeler.get_label_matrices(dev_cands)
+    print(L_dev[0].shape)
```
Comment (Wanille) [inline simple_model.py:64]:
```suggestion
    print("L_dev:", L_dev[0].shape)
```

File: simple_model.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+from fonduer.features import Featurizer
+from emmental.data import EmmentalDataLoader
+from fonduer.learning.dataset import FonduerDataset
+import numpy as np
+from emmental.model import EmmentalModel
+from fonduer.learning.task import create_task
+from emmental.learner import EmmentalLearner
+
+
+def main(db_name):
```
Comment (Wanille) [inline simple_model.py:27]:
```suggestion
def main(db_name: str):
```

File: doc_parser.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
+from pipeline.utils import get_session
+from fonduer.meta import Meta
+from fonduer.parser.preprocessors import HTMLDocPreprocessor
+from fonduer.parser import Parser
+
+
+def main(db_name):
```
Comment (Wanille) [inline doc_parser.py:7]:
```suggestion
def main(db_name: str):
```

File: doc_parser.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
+from pipeline.utils import get_session
+from fonduer.meta import Meta
+from fonduer.parser.preprocessors import HTMLDocPreprocessor
+from fonduer.parser import Parser
+
```
Comment (Wanille) [inline doc_parser.py:5]:
```suggestion
from pipeline.utils import PARALLEL

```

File: apply_lf.py [hunk 13]
```diff
@@ -0,0 +1,21 @@
+from fonduer.supervision import Labeler
+from fonduer.parser.models import Document
+
+from pipeline.utils import get_session
+from definitions.candidates import NameFullAbbr, NameFullTask
+from pipeline.lfs.dummy_lfs import dummy_lfs
+
```
Comment (Wanille) [inline apply_lf.py:7]:
```suggestion
from pipeline.utils import PARALLEL
```

File: simple_model.py [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+from definitions.candidates import NameFullAbbr, NameFullTask
+from pipeline.lfs.dummy_lfs import dummy_lfs
+
+from fonduer.candidates.models import Candidate
+from fonduer.candidates.matchers import LambdaFunctionMatcher, Union
+import re
+from snorkel.labeling import LFAnalysis # LFAnalysis creates statistic on Coverage, Overlap, Conflict
+from snorkel.labeling.model import LabelModel
+from pipeline.machine_learning.custom_get_candidates import custom_get_candidates
+
```
Comment (Wanille) [inline simple_model.py:14]:
```suggestion
from pipeline.utils import PARALLEL

```

File: simple_model.py [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+    test_cands = custom_get_candidates(session, 2, [NameFullAbbr, NameFullTask])
+
+    L_train = labeler.get_label_matrices(train_cands) # rows = candidates, columns = labelfuncs with -1, 0, 1
+    # ^-- these are needed for receiving the empirical accuracy as well as training the model
+
+    # (gold) labels must be loaded to calculate the empirical accuracy of our labeling functions with helpf of snorkel
+    # L_gold_train = labeler.get_gold_labels(train_cands, annotator="gold")
+
+    
+    LFAnalysis(
```
Comment (Wanille) [inline simple_model.py:47]:
```suggestion
lf_analysis = LFAnalysis(
```

File: simple_model.py [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+        L_train[0],
+        lfs=sorted(dummy_lfs, key=lambda lf: lf.name)
+    ).lf_summary() #Y=L_gold_train[0].reshape(-1)) #adds empirical accuracy to statistics
+
```
Comment (Wanille) [inline simple_model.py:51]:
```suggestion
print("LFAnalysis:", lf_analysis)

```

File: simple_model.py [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+        word_counter=word_counter, word_dim=300, specials=specials
+    )
+
+    # Needs to be done somehwhere before along with extracting the candidates maybe?:
+
+    featurizer = Featurizer(session, [NameFullAbbr, NameFullTask])
+    # featurizer.apply(split=0, train=True, parallelism=PARALLEL)
+    F_train = featurizer.get_feature_matrices(train_cands)
+
+    print(F_train[0].shape)
```
Comment (Wanille) [inline simple_model.py:123]:
```suggestion
    print("F_train:", F_train[0].shape)
```

File: simple_model.py [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+    
+    # featurizer.apply(split=1, parallelism=PARALLEL)
+    F_dev = featurizer.get_feature_matrices(dev_cands)
+    print(F_dev[0].shape)
```
Comment (Wanille) [inline simple_model.py:127]:
```suggestion
    print("F_dev:", F_dev[0].shape)
```

File: simple_model.py [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+
+    #featurizer.apply(split=2, parallelism=PARALLEL)
+    F_test = featurizer.get_feature_matrices(test_cands)
+    print(F_test[0].shape)
```
Comment (Wanille) [inline simple_model.py:131]:
```suggestion
    print("F_test:", F_test[0].shape)
```

File: simple_model.py [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+
+
+    # Generate dataloader for training set
+   
+    # Filter out noise samples
+    diffs = train_marginals.max(axis=1) - train_marginals.min(axis=1)
+    train_idxs = np.where(diffs > 1e-6)[0]
+
+    train_dataloader = EmmentalDataLoader(
+        task_to_label_dict={"test_collection": "labels"},
```
Comment (Wanille) [inline simple_model.py:141]:
```suggestion
        task_to_label_dict={db_name: "labels"},
```

File: simple_model.py [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+            "n_epochs": 30,
+            "optimizer_config": {"lr": 0.001, "l2": 0.0},
+            "task_scheduler": "round_robin",
+        },
+        "logging_config": {
+            "evaluation_freq": 1,
+            "counter_unit": "epoch",
+            "checkpointing": False,
+            "checkpointer_config": {
+                "checkpoint_metric": {f"test_collection/test_collection/train/loss": "min"},
```
Comment (Wanille) [inline simple_model.py:84]:
```suggestion
                "checkpoint_metric": {f"{db_name}/{db_name}/train/loss": "min"},
```

File: simple_model.py [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
+    # Generate dataloader for training set
+   
+    # Filter out noise samples
+    diffs = train_marginals.max(axis=1) - train_marginals.min(axis=1)
+    train_idxs = np.where(diffs > 1e-6)[0]
+
+    train_dataloader = EmmentalDataLoader(
+        task_to_label_dict={"test_collection": "labels"},
+        dataset=FonduerDataset(
+            "test_collection",
```
Comment (Wanille) [inline simple_model.py:143]:
```suggestion
            db_name,
```

File: simple_model.py [hunk 23]
```diff
@@ -0,0 +1,175 @@
...
+        ),
+        split="train",
+        batch_size=100,
+        shuffle=True,
+    )
+
+
+    # Create task and model, and perform training
+    tasks = create_task(
+        "test_collection", 2, F_train[0].shape[1], 2, emb_layer, model="LogisticRegression"
```
Comment (Wanille) [inline simple_model.py:158]:
```suggestion
        db_name, 2, F_train[0].shape[1], 2, emb_layer, model="LogisticRegression"
```

File: simple_model.py [hunk 24]
```diff
@@ -0,0 +1,175 @@
...
+    )
+
+    model = EmmentalModel(name=f"test_collection_task")
```
Comment (Wanille) [inline simple_model.py:161]:
```suggestion
    model = EmmentalModel(name=f"{db_name}_task")
```

**Review by Wanille — COMMENTED** [review]
Good stuff guys please read trough the comments and apply the changes if they make sense to you. Also look over the rest and see if I forgot some stuff and if comment your ideas :grin:

**Review by Pexem — APPROVED** [review]

**Pull request merged**